boards: arm: lpcxpresso51u68: Add LPCXpresso51U68 board definition

Navin-Sankar opened the pull request:
Add SoC and board definition for the NXP LPCXpresso51U68 development
board.

Added adc driver support for LPC51 series soc.

Signed-off-by: Navin Sankar Velliangiri <navin@linumiz.com>

Commit:
boards: arm: lpcxpresso51u68: Add LPCXpresso51U68 board definition

Add board definition for the NXP LPCXpresso51U68 development
board.

Signed-off-by: Navin Sankar Velliangiri <navin@linumiz.com>

galak commented on drivers/gpio/gpio_mcux_lpc.c:
```diff
@@ -230,11 +230,19 @@ static uint32_t attach_pin_to_isr(uint32_t port, uint32_t pin, uint32_t isr_no)
 	/* Code asumes PIN_INT values are grouped [0..3] and [4..7].
 	 * This scenario is true in LPC54xxx/LPC55xxx.
 	 */
+#ifdef CONFIG_BOARD_LPCXPRESSO51U68
```
We need to find a better way to handle this, we shouldn't introduce board specific ifdef's in this file.

galak replied:
Also, does this make sense on LPC51U68 in that it seems like only has PINT0/1/2/3 so the else case doesn't make any sense.

galak reviewed (changes requested):
Need to cleanup how we handle the PIN_INT differences for LPC51U68.

Navin-Sankar commented:
@galak can this tagged for 2.7?

dleach02 commented on drivers/gpio/gpio_mcux_lpc.c:
```diff
@@ -51,8 +51,13 @@ struct gpio_mcux_lpc_data {
+#ifdef CONFIG_SOC_SERIES_LPC51U68
 	/* ISR allocated in device tree to this port */
-	uint32_t isr_list[8];
+	uint32_t isr_list[4];
```
Instead of having conditional code based on the platform, there is a define in the SOC header file for the count here that would be cleaner:

INPUTMUX_PINTSEL_COUNT

Navin-Sankar replied:
@dleach02 fixed.

MaureenHelm commented on modules/Kconfig.mcux:
```diff
@@ -25,6 +25,11 @@ config HAS_MCUX_LPADC
 	help
 	  Set if the LPADC module is present in the SoC.
 
+config HAS_LPC_ADC
+	bool
+	help
+	  Set if the LPC ADC module is present in the SoC.
+
```
This change should be moved to the commit `drivers: adc: added support adc drive for lpc51u68` or a new commit

MaureenHelm commented on soc/arm/nxp_lpc/lpc51u68/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,42 @@
+	default y
+	depends on PWM
+
+config HAS_LPC_ADC
```
This should be `ADC_LPC_ADC`, and added in a later commit after that symbol is introduced

MaureenHelm reviewed (dismissed):
Thank you for the contribution!

Can you please take a look at the buildkite failures?

Navin-Sankar commented:
@dleach02 addressed the conflicts and change requests.
The [hal](https://github.com/zephyrproject-rtos/hal_nxp/pull/95) layer need to be addressed and merged before this pr.

Navin-Sankar commented:
@dleach02 CI failure is related to this [pull request](https://github.com/zephyrproject-rtos/hal_nxp/pull/113)

dleach02 commented:
> @dleach02 CI failure is related to this [pull request](https://github.com/zephyrproject-rtos/hal_nxp/pull/113)

Yeah, we found out that the LPC and RT6xx had different names for this define. Looking to address this in a patch to the HAL.

parthitce commented:
Fixes tag in https://github.com/zephyrproject-rtos/hal_nxp/pull/113 closed it. Re-opening.

Navin-Sankar commented:
@dleach02 Please revisit again

dleach02 commented:
@parthitce, the HAL PR has been merged so you need to resolve the west.yml conflicts

dleach02 commented:
@parthitce, can you rebase this PR and verify the functionality? 

I believe you should be able to just use the west.yml in main now.

dleach02 commented:
@galak, @MaureenHelm can you two confirm if your blocks are still valid on this PR? Thank you.

Navin-Sankar commented on drivers/adc/adc_lpc_adc.c:
```diff
@@ -0,0 +1,259 @@
+	adc_config_t adc_config;
+	uint32_t frequency = 0;
+
+	POWER_DisablePD(kPDRUNCFG_PD_ADC0);
+	POWER_DisablePD(kPDRUNCFG_PD_VD7_ENA);
+	POWER_DisablePD(kPDRUNCFG_PD_VREFP_SW);
+	POWER_DisablePD(kPDRUNCFG_PD_TEMPS);
```
@mmahadevan108 any updates on this. where to place the adc power related control's?
Either in driver file or soc.c file.

mmahadevan108 replied:
Agree with @MaureenHelm, this should not be in the driver. Probably `soc_init` is the place for this

mmahadevan108 replied:
@Navin-Sankar, could kindly take a look at addressing this comment.

mmahadevan108 commented on dts/arm/nxp/nxp_lpc51u68.dtsi:
```diff
@@ -0,0 +1,193 @@
+
+&nvic {
+	arm,num-irq-priority-bits = <2>;
+};
```
Please move this dts file to a separate commit.

mmahadevan108 commented on soc/arm/nxp_lpc/lpc51u68/linker.ld:
```diff
@@ -0,0 +1,9 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <autoconf.h>
```
This include can be deleted

mmahadevan108 commented on soc/arm/nxp_lpc/lpc51u68/soc.h:
```diff
@@ -0,0 +1,34 @@
+#ifndef _ASMLANGUAGE
+#include <sys/util.h>
+
+#include <devicetree.h>
```
This include can be deleted

mmahadevan108 commented on drivers/adc/adc_lpc_adc.c:
```diff
@@ -0,0 +1,259 @@
+	CLOCK_EnableClock(kCLOCK_Adc0);
+
+	base->CTRL |= ADC_CTRL_BYPASSCAL_MASK;
+	frequency = CLOCK_GetFreq(kCLOCK_BusClk);
```
This should be using the `clock_control` driver API. You would need to add device tree properties for the adc. Please refer the LPC UART driver.

Navin-Sankar replied:
I didn't get your point. How to define clock frequency in dts file?

mmahadevan108 replied:
Below is an example of how we get clock frequency for the I2C driver.
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_mcux_flexcomm.c#L65

Below is an example of how we specify in the device tree which clock
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/nxp/nxp_lpc55S1x_common.dtsi#L150

Below is the clock control driver for LPC:
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/clock_control/clock_control_mcux_syscon.c

Hope this helps.

Navin-Sankar replied:
Thanks for the hint.

Navin-Sankar replied:
@mmahadevan108 the current **clock_control** driver support's only to get the frequency for peripheral clocks like flexcomm, can, timers.
But for this adc driver, **BusClock** is used to get the clock frequency. 

For that we need to introduce a new macro in a header file **dt-bindings/clock/mcux_lpc_syscon_clock.h**  to get the clock frequency of system clock's.

Is this approach is correct?

mmahadevan108 replied:
Yes, that would be the approach to take.

Commit:
dts: LPC51U68 Initial device tree

Create device tree for LPC51U68 MCU

Signed-off-by: Navin Sankar Velliangiri <navin@linumiz.com>

Commit:
drivers: hwinfo: Exclude SYSCON driver on LPC51U68

The hwinfo_mcux_syscon driver isn't relavant for LPC51U68 SoC
series. Add a Kconfig exclusion so the driver isn't available
on LPC51U68 SoC.

Signed-off-by: Navin Sankar Velliangiri <navin@linumiz.com>

Navin-Sankar commented:
@mmahadevan108 Added busclock support [here](https://github.com/zephyrproject-rtos/zephyr/pull/43554)

mmahadevan108 commented:
Would you like to rebase and update

dleach02 commented:
@galak can you review to determine if your change request has been addressed?

Commit:
soc: arm: nxp: Add LPC51U68 support.

Add initial support for NXP LPC51U68 SOC series

Signed-off-by: Navin Sankar Velliangiri <navin@linumiz.com>

dleach02 commented:
@MaureenHelm @galak, can you revisit your  change requests to determine if @Navin-Sankar has addressed your items?

mmahadevan108 commented on boards/arm/lpcxpresso51u68/lpcxpresso51u68.dts:
```diff
@@ -0,0 +1,105 @@
+	pwmleds {
+		compatible = "pwm-leds";
+		green_pwm_led: green_pwm_led {
+			pwms = <&sc_timer 4 PWM_POLARITY_NORMAL>;
```
This might need to be updated per changes to PWM. This might be causing the CI failures.

dleach02 reviewed (changes requested):
@Navin-Sankar, this will need to be transitioned to using pinctrl

dleach02 commented:
@Navin-Sankar do you plan on revisiting this PR?

Navin-Sankar commented:
Currently working on this, requested a help from @danieldegrasse to generate the pincontrol file.

dleach02 commented:
@Navin-Sankar you will need to update the west.yml to point to the correct NXP HAL sha that merged your pinctrl changes.

galak commented on soc/arm/nxp_lpc/lpc51u68/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,38 @@
+	default y
+	depends on FLASH
+
+config PWM_MCUX_SCTIMER
```
remove PWM_MCUX_SCTIMER, not needed as devicetree will enable this.

galak commented on soc/arm/nxp_lpc/lpc51u68/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,38 @@
+	default y
+	depends on GPIO
+
+config UART_MCUX_FLEXCOMM
```
remove UART_MCUX_FLEXCOMM as devicetree will enable this.

galak commented on soc/arm/nxp_lpc/lpc51u68/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,38 @@
+	default y
+	depends on PINMUX
+
+config GPIO_MCUX_LPC
```
remove GPIO_MCUX_LPC as devicetree will enable this

galak commented on soc/arm/nxp_lpc/lpc51u68/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,38 @@
+config SOC
+	default "lpc51u68"
+
+config PINMUX_MCUX_LPC
```
Is pinmux needed?

galak commented on drivers/hwinfo/Kconfig:
```diff
@@ -67,7 +67,8 @@ config HWINFO_MCUX_SRC_V2
 	default y
-	depends on HAS_MCUX_SYSCON && !SOC_SERIES_LPC54XXX
+	depends on HAS_MCUX_SYSCON && !SOC_SERIES_LPC54XXX \
+		   && !SOC_SERIES_LPC51U68
```
Change this to `depends on DT_HAS_NXP_LPC_UID`

galak reviewed (dismissed):
Some Kconfig cleanups needed

mmahadevan108 commented:
@Navin-Sankar. Please rebase and remove your changes to `west.yml`

galak commented on dts/arm/nxp/nxp_lpc51u68.dtsi:
```diff
@@ -0,0 +1,173 @@
+		syscon: syscon@40000000 {
+			compatible = "nxp,lpc-syscon";
+			reg = <0x40000000 0x4000>;
+			label = "SYSCON";
```
remove `label`

Navin-Sankar commented on dts/arm/nxp/nxp_lpc51u68.dtsi:
```diff
@@ -0,0 +1,173 @@
+		flash0: flash@0 {
+			compatible = "soc-nv-flash";
+			reg = <0 DT_SIZE_K(256)>;
+			label = "LPC_FLASH";
```
done

mmahadevan108 commented on drivers/gpio/gpio_mcux_lpc.c:
```diff
@@ -265,6 +265,9 @@ static uint32_t attach_pin_to_isr(uint32_t port, uint32_t pin, uint32_t isr_no)
 #elif (FSL_FEATURE_PINT_NUMBER_OF_CONNECTED_OUTPUTS > 4)
+#ifdef CONFIG_SOC_SERIES_LPC51U68
+	pint_idx = isr_no - PIN_INT0_IRQn;
+#endif
```
Please delete the above 3 lines as LPC51U68 has less than 4 outputs and should never go into this block of code.

mmahadevan108 reviewed (changes requested):
Please update the changes to the GPIO driver

Commit:
drivers: gpio: gpio_mcux_lpc: Allocate no of gpio ISR using macro

Allocate no of gpio ISR using macro INPUTMUX_PINTSEL_COUNT.

Signed-off-by: Navin Sankar Velliangiri <navin@linumiz.com>
